File: interop/python/test_reflection.py
#!/usr/bin/env python3
"""
Comprehensive test suite for Flow Python bindings reflection API
Tests the ability to introspect Flow modules from Python
"""

import sys
import os
sys.path.insert(0, os.path.dirname(__file__))

import flow

# Test counters
tests_passed = 0
tests_failed = 0

def assert_equal(actual, expected, test_name):
    global tests_passed, tests_failed
    if actual == expected:
        print(f"✓ {test_name}")
        tests_passed += 1
    else:
        print(f"✗ {test_name}")
        print(f"  Expected: {expected}")
        print(f"  Got: {actual}")
        tests_failed += 1

def assert_true(condition, test_name):
    assert_equal(condition, True, test_name)

def assert_in(item, container, test_name):
    global tests_passed, tests_failed
    if item in container:
        print(f"✓ {test_name}")
        tests_passed += 1
    else:
        print(f"✗ {test_name}")
        print(f"  Expected '{item}' to be in {container}")
        tests_failed += 1

def assert_not_none(value, test_name):
    global tests_passed, tests_failed
    if value is not None:
        print(f"✓ {test_name}")
        tests_passed += 1
    else:
        print(f"✗ {test_name}")
        print(f"  Expected value to not be None")
        tests_failed += 1

print("=" * 60)
print("Flow Python Reflection API Tests")
print("=" * 60)

# Test 1: Initialize runtime
print("\n[Test 1] Initialize Runtime")
try:
    runtime = flow.Runtime()
    print("✓ Runtime initialized")
    tests_passed += 1
except Exception as e:
    print(f"✗ Runtime initialization failed: {e}")
    tests_failed += 1
    sys.exit(1)

# Test 2: Compile module with functions
print("\n[Test 2] Compile Module with Multiple Functions")
code = """
func add(a: int, b: int) -> int {
    return a + b;
}

func subtract(x: int, y: int) -> int {
    return x - y;
}

func multiply(a: int, b: int) -> int {
    return a * b;
}

func greet(name: string) -> string {
    return "Hello, " + name;
}

func get_pi() -> float {
    return 3.14159;
}
"""

try:
    module = flow.Module.compile(runtime, code)
    print("✓ Module compiled successfully")
    tests_passed += 1
except Exception as e:
    print(f"✗ Module compilation failed: {e}")
    tests_failed += 1
    sys.exit(1)

# Test 3: Get function count
print("\n[Test 3] Get Function Count")
try:
    count = module.get_function_count()
    assert_equal(count, 5, "Function count should be 5")
except Exception as e:
    print(f"✗ get_function_count failed: {e}")
    tests_failed += 1

# Test 4: List all functions
print("\n[Test 4] List All Functions")
try:
    functions = module.list_functions()
    assert_equal(len(functions), 5, "Should have 5 function names")
    assert_in("add", functions, "'add' should be in function list")
    assert_in("subtract", functions, "'subtract' should be in function list")
    assert_in("multiply", functions, "'multiply' should be in function list")
    assert_in("greet", functions, "'greet' should be in function list")
    assert_in("get_pi", functions, "'get_pi' should be in function list")
except Exception as e:
    print(f"✗ list_functions failed: {e}")
    tests_failed += 1

# Test 5: Get function info for 'add'
print("\n[Test 5] Get Function Info - 'add'")
try:
    info = module.get_function_info("add")
    assert_not_none(info, "Function info should not be None")
    assert_equal(info['name'], "add", "Function name should be 'add'")
    assert_equal(info['return_type'], "int", "Return type should be 'int'")
    assert_equal(len(info['parameters']), 2, "Should have 2 parameters")
    assert_equal(info['parameters'][0]['name'], "a", "First param name should be 'a'")
    assert_equal(info['parameters'][0]['type'], "int", "First param type should be 'int'")
    assert_equal(info['parameters'][1]['name'], "b", "Second param name should be 'b'")
    assert_equal(info['parameters'][1]['type'], "int", "Second param type should be 'int'")
except Exception as e:
    print(f"✗ get_function_info('add') failed: {e}")
    tests_failed += 1

# Test 6: Get function info for 'greet' (string parameter)
print("\n[Test 6] Get Function Info - 'greet' (string parameter)")
try:
    info = module.get_function_info("greet")
    assert_equal(info['name'], "greet", "Function name should be 'greet'")
    assert_equal(info['return_type'], "string", "Return type should be 'string'")
    assert_equal(len(info['parameters']), 1, "Should have 1 parameter")
    assert_equal(info['parameters'][0]['name'], "name", "Param name should be 'name'")
    assert_equal(info['parameters'][0]['type'], "string", "Param type should be 'string'")
except Exception as e:
    print(f"✗ get_function_info('greet') failed: {e}")
    tests_failed += 1

# Test 7: Get function info for 'get_pi' (no parameters)
print("\n[Test 7] Get Function Info - 'get_pi' (no parameters)")
try:
    info = module.get_function_info("get_pi")
    assert_equal(info['name'], "get_pi", "Function name should be 'get_pi'")
    assert_equal(info['return_type'], "float", "Return type should be 'float'")
    assert_equal(len(info['parameters']), 0, "Should have 0 parameters")
except Exception as e:
    print(f"✗ get_function_info('get_pi') failed: {e}")
    tests_failed += 1

# Test 8: Error handling - non-existent function
print("\n[Test 8] Error Handling - Non-existent Function")
try:
    info = module.get_function_info("nonexistent")
    print(f"✗ Should have raised FlowError for non-existent function")
    tests_failed += 1
except flow.FlowError as e:
    print(f"✓ Correctly raised FlowError: {e}")
    tests_passed += 1
except Exception as e:
    print(f"✗ Wrong exception type: {e}")
    tests_failed += 1

# Test 9: Module inspect() method
print("\n[Test 9] Module inspect() Method")
try:
    inspection = module.inspect()
    assert_true("add(a: int, b: int) -> int" in inspection, "Should show 'add' signature")
    assert_true("greet(name: string) -> string" in inspection, "Should show 'greet' signature")
    assert_true("get_pi() -> float" in inspection, "Should show 'get_pi' signature")
    print(f"\nInspection output:\n{inspection}")
except Exception as e:
    print(f"✗ inspect() failed: {e}")
    tests_failed += 1

# Test 10: Module with boolean functions
print("\n[Test 10] Boolean Return Types")
code_bool = """
func is_positive(x: int) -> bool {
    return x > 0;
}

func is_even(n: int) -> bool {
    return n % 2 == 0;
}
"""
try:
    module_bool = flow.Module.compile(runtime, code_bool)
    info = module_bool.get_function_info("is_positive")
    assert_equal(info['return_type'], "bool", "Return type should be 'bool'")
    
    functions = module_bool.list_functions()
    assert_equal(len(functions), 2, "Should have 2 functions")
except Exception as e:
    print(f"✗ Boolean functions test failed: {e}")
    tests_failed += 1

# Test 11: Large module with many functions
print("\n[Test 11] Large Module with Many Functions")
code_large = "\n".join([
    f"func func{i}(x: int) -> int {{ return x * {i}; }}"
    for i in range(20)
])
try:
    module_large = flow.Module.compile(runtime, code_large)
    count = module_large.get_function_count()
    assert_equal(count, 20, "Should have 20 functions")
    
    functions = module_large.list_functions()
    assert_equal(len(functions), 20, "Should list 20 functions")
    
    # Check a few function infos
    info0 = module_large.get_function_info("func0")
    assert_equal(info0['name'], "func0", "Function 0 name")
    
    info19 = module_large.get_function_info("func19")
    assert_equal(info19['name'], "func19", "Function 19 name")
except Exception as e:
    print(f"✗ Large module test failed: {e}")
    tests_failed += 1

# Test 12: Verify functions still callable after reflection
print("\n[Test 12] Functions Still Callable After Reflection")
try:
    # Use the original module
    result = module.call("add", 10, 20)
    assert_equal(result, 30, "add(10, 20) should return 30")
    
    result = module.call("multiply", 7, 6)
    assert_equal(result, 42, "multiply(7, 6) should return 42")
except Exception as e:
    print(f"✗ Calling functions after reflection failed: {e}")
    tests_failed += 1

# Test 13: Multiple parameter types
print("\n[Test 13] Mixed Parameter Types")
code_mixed = """
func process(a: int, b: float, c: string, d: bool) -> string {
    return "processed";
}
"""
try:
    module_mixed = flow.Module.compile(runtime, code_mixed)
    info = module_mixed.get_function_info("process")
    
    assert_equal(len(info['parameters']), 4, "Should have 4 parameters")
    assert_equal(info['parameters'][0]['type'], "int", "Param 0 type")
    assert_equal(info['parameters'][1]['type'], "float", "Param 1 type")
    assert_equal(info['parameters'][2]['type'], "string", "Param 2 type")
    assert_equal(info['parameters'][3]['type'], "bool", "Param 3 type")
except Exception as e:
    print(f"✗ Mixed parameter types test failed: {e}")
    tests_failed += 1

# Test 14: Empty module
print("\n[Test 14] Empty Module")
try:
    module_empty = flow.Module.compile(runtime, "// Just a comment")
    count = module_empty.get_function_count()
    assert_equal(count, 0, "Empty module should have 0 functions")
    
    functions = module_empty.list_functions()
    assert_equal(len(functions), 0, "Empty module should list 0 functions")
    
    inspection = module_empty.inspect()
    assert_true("no functions" in inspection.lower(), "Inspection should mention no functions")
except Exception as e:
    print(f"✗ Empty module test failed: {e}")
    tests_failed += 1

# Test 15: Reflection on loaded file module
print("\n[Test 15] Reflection on File-Loaded Module")
test_file = "/tmp/test_reflect.flow"
with open(test_file, "w") as f:
    f.write("""
func square(x: int) -> int {
    return x * x;
}

func cube(x: int) -> int {
    return x * x * x;
}
""")

try:
    module_file = flow.Module.load(runtime, test_file)
    functions = module_file.list_functions()
    assert_in("square", functions, "'square' should be in loaded module")
    assert_in("cube", functions, "'cube' should be in loaded module")
    
    info = module_file.get_function_info("square")
    assert_equal(info['name'], "square", "Loaded function name")
except Exception as e:
    print(f"✗ File-loaded module reflection failed: {e}")
    tests_failed += 1
finally:
    if os.path.exists(test_file):
        os.remove(test_file)

# Summary
print("\n" + "=" * 60)
print(f"Tests Passed: {tests_passed}")
print(f"Tests Failed: {tests_failed}")
print(f"Total Tests: {tests_passed + tests_failed}")
print("=" * 60)

if tests_failed == 0:
    print("✓ All tests passed!")
    sys.exit(0)
else:
    print(f"✗ {tests_failed} test(s) failed")
    sys.exit(1)

File: interop/php/test_reflection.php
#!/usr/bin/env php
<?php


require_once __DIR__ . '/src/Flow.php';

use Flow\Flow;
use Flow\FlowException;

$testsPassed = 0;
$testsFailed = 0;

function test(string $name, callable $fn): void
{
    global $testsPassed, $testsFailed;

    try {
        $fn();
        echo "✓ $name\n";
        $testsPassed++;
    } catch (Exception $e) {
        echo "✗ $name\n";
        echo "  {$e->getMessage()}\n";
        $testsFailed++;
    }
}

function assertEqual($actual, $expected, string $message): void
{
    if ($actual !== $expected) {
        throw new Exception("$message: expected " . var_export($expected, true) . ", got " . var_export($actual, true));
    }
}

function assertArrayEqual(array $actual, array $expected, string $message): void
{
    sort($actual);
    sort($expected);
    if ($actual !== $expected) {
        throw new Exception("$message: expected " . json_encode($expected) . ", got " . json_encode($actual));
    }
}

function assertTrue(bool $condition, string $message): void
{
    if (!$condition) {
        throw new Exception($message);
    }
}

function assertContains(string $haystack, string $needle, string $message): void
{
    if (strpos($haystack, $needle) === false) {
        throw new Exception("$message: '$haystack' does not contain '$needle'");
    }
}

echo "============================================================\n";
echo "Flow PHP Reflection API Tests\n";
echo "============================================================\n\n";


test('test_function_count', function() {
    $code = <<<'FLOW'
func add(a: int, b: int) -> int {
    return a + b;
}

func subtract(x: int, y: int) -> int {
    return x - y;
}

func multiply(m: int, n: int) -> int {
    return m * n;
}
FLOW;

    $module = Flow::compile($code);
    $count = $module->functionCount();

    assertEqual($count, 3, 'Should have 3 functions');
});


test('test_list_functions', function() {
    $code = <<<'FLOW'
func greet(name: string) -> string {
    return "Hello, " + name;
}

func square(x: int) -> int {
    return x * x;
}

func is_positive(n: int) -> bool {
    return n > 0;
}
FLOW;

    $module = Flow::compile($code);
    $functions = $module->listFunctions();

    assertEqual(count($functions), 3, 'Should have 3 function names');
    assertArrayEqual($functions, ['greet', 'square', 'is_positive'], 'Function names should match');
});


test('test_function_info_with_parameters', function() {
    $code = <<<'FLOW'
func add(a: int, b: int) -> int {
    return a + b;
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('add');

    assertEqual($info['name'], 'add', 'Function name should be add');
    assertEqual($info['return_type'], 'int', 'Return type should be int');
    assertEqual(count($info['parameters']), 2, 'Should have 2 parameters');
    assertEqual($info['parameters'][0]['name'], 'a', 'First param name');
    assertEqual($info['parameters'][0]['type'], 'int', 'First param type');
    assertEqual($info['parameters'][1]['name'], 'b', 'Second param name');
    assertEqual($info['parameters'][1]['type'], 'int', 'Second param type');
});


test('test_function_info_string_parameter', function() {
    $code = <<<'FLOW'
func greet(name: string) -> string {
    return "Hello, " + name;
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('greet');

    assertEqual($info['name'], 'greet', 'Function name');
    assertEqual($info['return_type'], 'string', 'Return type');
    assertEqual(count($info['parameters']), 1, 'Parameter count');
    assertEqual($info['parameters'][0]['name'], 'name', 'Param name');
    assertEqual($info['parameters'][0]['type'], 'string', 'Param type');
});


test('test_function_info_no_parameters', function() {
    $code = <<<'FLOW'
func get_pi() -> float {
    return 3.14159;
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('get_pi');

    assertEqual($info['name'], 'get_pi', 'Function name');
    assertEqual($info['return_type'], 'float', 'Return type');
    assertEqual(count($info['parameters']), 0, 'Should have no parameters');
});


test('test_function_info_nonexistent', function() {
    $code = <<<'FLOW'
func test() -> int {
    return 42;
}
FLOW;

    $module = Flow::compile($code);

    $threw = false;
    try {
        $module->functionInfo('nonexistent');
    } catch (FlowException $e) {
        $threw = true;
        assertContains($e->getMessage(), 'not found', 'Error message should mention not found');
    }

    assertTrue($threw, 'Should throw error for non-existent function');
});


test('test_inspect_functions', function() {
    $code = <<<'FLOW'
func add(a: int, b: int) -> int {
    return a + b;
}

func greet(name: string) -> string {
    return "Hello";
}
FLOW;

    $module = Flow::compile($code);
    $inspection = $module->inspect();

    assertContains($inspection, 'add(a: int, b: int) -> int', 'Should contain add signature');
    assertContains($inspection, 'greet(name: string) -> string', 'Should contain greet signature');
    assertContains($inspection, '2 function(s)', 'Should mention 2 functions');
});


test('test_boolean_return_types', function() {
    $code = <<<'FLOW'
func is_positive(x: int) -> bool {
    return x > 0;
}

func is_even(n: int) -> bool {
    return n % 2 == 0;
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('is_positive');

    assertEqual($info['return_type'], 'bool', 'Return type should be bool');
    assertEqual($module->functionCount(), 2, 'Should have 2 functions');
});


test('test_large_module', function() {
    $code = '';
    for ($i = 0; $i < 20; $i++) {
        $code .= "func func$i(x: int) -> int { return x * $i; }\n";
    }

    $module = Flow::compile($code);
    $count = $module->functionCount();

    assertEqual($count, 20, 'Should have 20 functions');

    $functions = $module->listFunctions();
    assertEqual(count($functions), 20, 'Should list 20 functions');

    $info0 = $module->functionInfo('func0');
    assertEqual($info0['name'], 'func0', 'First function name');

    $info19 = $module->functionInfo('func19');
    assertEqual($info19['name'], 'func19', 'Last function name');
});


test('test_functions_still_callable_after_reflection', function() {
    $code = <<<'FLOW'
func add(a: int, b: int) -> int {
    return a + b;
}

func multiply(x: int, y: int) -> int {
    return x * y;
}
FLOW;

    $module = Flow::compile($code);

    // Do reflection
    $count = $module->functionCount();
    assertEqual($count, 2, 'Should have 2 functions');

    $info = $module->functionInfo('add');
    assertEqual($info['name'], 'add', 'Function name should be add');

    // Now call the functions to ensure reflection didn't break them
    $result1 = $module->call('add', 10, 20);
    assertEqual($result1, 30, 'add(10, 20) should return 30');

    $result2 = $module->call('multiply', 7, 6);
    assertEqual($result2, 42, 'multiply(7, 6) should return 42');
});

// Test 11: Mixed parameter types
test('test_mixed_parameter_types', function() {
    $code = <<<'FLOW'
func process(a: int, b: float, c: string, d: bool) -> string {
    return "processed";
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('process');

    assertEqual(count($info['parameters']), 4, 'Should have 4 parameters');
    assertEqual($info['parameters'][0]['type'], 'int', 'First param type');
    assertEqual($info['parameters'][1]['type'], 'float', 'Second param type');
    assertEqual($info['parameters'][2]['type'], 'string', 'Third param type');
    assertEqual($info['parameters'][3]['type'], 'bool', 'Fourth param type');
});

// Test 12: Empty module
test('test_empty_module', function() {
    $code = '// Just a comment';

    $module = Flow::compile($code);
    $count = $module->functionCount();

    assertEqual($count, 0, 'Empty module should have 0 functions');

    $functions = $module->listFunctions();
    assertEqual(count($functions), 0, 'Empty module should list 0 functions');

    $inspection = $module->inspect();
    assertContains(strtolower($inspection), 'no functions', 'Should mention no functions');
});

// Test 13: Float return type
test('test_float_return_type', function() {
    $code = <<<'FLOW'
func calculate_pi() -> float {
    return 3.14159;
}
FLOW;

    $module = Flow::compile($code);
    $info = $module->functionInfo('calculate_pi');

    assertEqual($info['return_type'], 'float', 'Return type should be float');
    assertEqual(count($info['parameters']), 0, 'Should have no parameters');
});

// Test 14: Multiple string functions
test('test_multiple_string_functions', function() {
    $code = <<<'FLOW'
func concat(a: string, b: string) -> string {
    return a + b;
}

func repeat(s: string, n: int) -> string {
    return s;
}
FLOW;

    $module = Flow::compile($code);

    $info1 = $module->functionInfo('concat');
    assertEqual(count($info1['parameters']), 2, 'concat should have 2 parameters');
    assertEqual($info1['parameters'][0]['type'], 'string', 'First param type');
    assertEqual($info1['parameters'][1]['type'], 'string', 'Second param type');

    $info2 = $module->functionInfo('repeat');
    assertEqual(count($info2['parameters']), 2, 'repeat should have 2 parameters');
    assertEqual($info2['parameters'][0]['type'], 'string', 'First param type');
    assertEqual($info2['parameters'][1]['type'], 'int', 'Second param type');
});

// Print summary
echo "\n============================================================\n";
echo "Tests passed: $testsPassed\n";
echo "Tests failed: $testsFailed\n";
echo "============================================================\n";

exit($testsFailed > 0 ? 1 : 0);

